In [1]:
import acids_and_bases as ab
from compound import Compound
import math

# Describe the relative acidity or basicity of a solution

In [2]:
# --- Given values ---
oh_conc = 1e-1  # M

# --- Analysis ---
solution_description = ab.describe_solution_from_oh(oh_conc)

print(solution_description)

The pOH is 1.0 and the pH is 13.0. The solution is highly basic.


# Calculate pH based on ion concentration

In [3]:
# --- Given values ---
h_ion_concentration = 7.9e-6  # M

# --- Calculation ---
pH_value = ab.calculate_pH(h_ion_concentration)

print(f"The pH of the solution is {pH_value:.2f}")


The pH of the solution is 5.10


# Salt is acidic

In [4]:
# --- Salts from the question ---
# Note: The formulas in the image had typos. Using correct formulas.
salts_to_test = {
    "Potassium Formate": "KCHO2", # Corrected from KCO₂H
    "Potassium Bromide": "KBr",
    "Ammonium Bromide": "NH4Br"
}

# --- Analysis ---
print("--- Salt Acidity Analysis ---")
for name, formula in salts_to_test.items():
    acidity = ab.predict_salt_acidity(formula)
    print(f"{name} ({formula}): {acidity}")


--- Salt Acidity Analysis ---
Potassium Formate (KCHO2): Basic
Potassium Bromide (KBr): Neutral
Ammonium Bromide (NH4Br): Acidic


# Rank oxyacids

In [5]:
# --- Acids to compare ---
oxyacids = ["HBrO", "HIO", "HClO"]

# --- Get the ranking ---
ranking = ab.rank_oxyacids_by_strength(oxyacids)

print("The order of oxyacids from strongest to weakest is:")
print(ranking)


The order of oxyacids from strongest to weakest is:
HClO > HBrO > HIO


In [6]:
# --- Acids to compare ---
oxyacids = ["HC2ClH2O2", "HC2BrH2O2", "HC2FH2O2"]

# --- Get the ranking ---
ranking = ab.rank_oxyacids_by_strength(oxyacids)

print("The order of oxyacids from strongest to weakest is:")
print(ranking)


Could not process HC2ClH2O2: Electronegativity for 'C2Cl22' not found.
Could not process HC2BrH2O2: Electronegativity for 'C2Br22' not found.
Could not process HC2FH2O2: Electronegativity for 'C2F22' not found.
The order of oxyacids from strongest to weakest is:



# hydroxide ion concentration based on concentration of h+

In [7]:
# --- Given values ---
h_conc = 1e-2  # M

# --- Calculation ---
oh_conc = ab.calculate_hydroxide_from_hydronium(h_conc)

print(f"The hydroxide ion concentration is {oh_conc:.0e} M")


The hydroxide ion concentration is 1e-12 M


# hydroxide ion concentration based on pOH

In [8]:
# --- Given values ---
pOH_value = 5.60

# --- Calculation ---
hydroxide_concentration = ab.calculate_hydroxide_from_pOH(pOH_value)

print(f"The hydroxide ion concentration is {hydroxide_concentration:.1e} M")

The hydroxide ion concentration is 2.5e-06 M


# Weak Acid Equilibrium (ICE Table)

## pH

In [9]:
# --- Given values ---
initial_concentration = 0.0480 # M
Ka_carbonic_acid = 4.30e-7

# --- Calculation ---
pH = ab.calculate_weak_acid_ph(initial_concentration, Ka_carbonic_acid)

print(f"The pH of the solution is {pH:.1f}")

The pH of the solution is 3.8


## pOH

In [10]:
# --- Given values ---
initial_concentration = 0.31 # M
Kb_ammonia = 1.8e-5

# --- Calculation ---
pOH = ab.calculate_weak_base_pOH(initial_concentration, Kb_ammonia)

print(f"The pOH of the solution is {pOH:.1f}")


The pOH of the solution is 2.6


## Concentration

In [11]:
# --- Given values ---
initial_conc_h2s = 0.0250 # M
Ka_h2s = 1.30e-7

# --- Calculation ---
eq_concentrations = ab.calculate_weak_acid_dissociation(initial_conc_h2s, Ka_h2s)

# The concentration of HS⁻ is the value for the conjugate base, A⁻
hs_concentration = eq_concentrations["A-"]

print(f"The equilibrium concentration of HS⁻ is {hs_concentration:.2e} M")

The equilibrium concentration of HS⁻ is 5.70e-05 M


# Convert K_b to K_a

In [12]:
# --- Given values ---
# This is the Kb for the lactate ion (conjugate base of lactic acid)
kb_lactate = 7.1e-11

# --- Calculation ---
# We calculate the Ka of its conjugate acid (lactic acid)
ka_lactic_acid = ab.convert_ka_kb(kb_lactate)

print(f"The Ka for lactic acid is {ka_lactic_acid:.1e}")

The Ka for lactic acid is 1.4e-04


# Convert K_a to K_b

In [13]:
# --- Given values ---
ka_formic_acid = 1.8e-4

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 5.6e-11


In [14]:
# --- Given values ---
ka_formic_acid = 6.2e-10

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 1.6e-05


In [15]:
# --- Given values ---
ka_formic_acid = 7.1e-10

# --- Calculation ---
# We calculate the Kb of its conjugate base (formate)
kb_formate = ab.convert_ka_kb(ka_formic_acid)

print(f"The Kb for the formate ion is {kb_formate:.1e}")

The Kb for the formate ion is 1.4e-05


# Molarity of a solution titration

In [16]:
# --- Given values ---
acid = "HNO3"
base = "Ca(OH)2"

# --- Calculation ---
# We are solving for the Molarity of the Acid (Ma)
result = ab.solve_titration(
    Ma=None,      # The unknown value
    Va_mL=25.0,
    na=1,         # HNO3 donates 1 proton
    Mb=0.175,
    Vb_mL=25.0,
    nb=2          # Ca(OH)2 accepts 2 protons
)

conc_acid = result['Ma'] # Extract the result from the dictionary
print(f"The molarity of the {acid} solution is {conc_acid:.3f} M")

The molarity of the HNO3 solution is 0.350 M


In [17]:
# --- Given values ---
acid = "H2SO4"
base = "LiOH"

# --- Calculation ---
# We are solving for the Molarity of the Base (Mb)
result = ab.solve_titration(
    Ma=3.2,
    Va_mL=13.1,
    na=2,         # H2SO4 donates 2 protons
    Mb=None,      # The unknown value
    Vb_mL=35,
    nb=1          # LiOH accepts 1 proton
)

conc_base = result['Mb'] # Extract the result
print(f"The concentration of the {base} solution is {conc_base:.1f} M")

The concentration of the LiOH solution is 2.4 M


In [18]:
# --- Given values ---
acid = "HCl"
base = "KOH"

# --- Calculation ---
# We are solving for the Molarity of the Acid (Ma)
result = ab.solve_titration(
    Ma=None,      # The unknown value
    Va_mL=20.0,
    na=1,         # HCl donates 1 proton
    Mb=1.15,
    Vb_mL=45.8,
    nb=1          # KOH accepts 1 proton
)

conc_acid = result['Ma'] # Extract the result
print(f"The concentration of the {acid} solution is {conc_acid:.2f} M")

The concentration of the HCl solution is 2.63 M


In [19]:
# Solve for the volume of the acid (Va_mL)
result = ab.solve_titration(
    Ma=1.25,      # Molarity of HCl
    Va_mL=None,   # We want to find this value
    na=1,         # HCl donates 1 proton
    Mb=1.75,      # Molarity of the base
    Vb_mL=12.58,  # Volume of the base
    nb=1          # For the FIRST equivalence point, the base accepts 1 proton
)

acid_volume = result['Va_mL']
print(f"The required volume of HCl is {acid_volume:.1f} mL")

The required volume of HCl is 17.6 mL


# [H3O+] of a solution

In [20]:
# --- Given values ---
Ka_HF = 6.3e-4
conc_HF = 0.500
conc_NaF = 0.750

# --- Calculation ---
h3o_concentration = ab.calculate_buffer_h3o_conc(Ka_HF, conc_HF, conc_NaF)

print(f"The [H₃O⁺] of the solution is {h3o_concentration:.1e} M")


The [H₃O⁺] of the solution is 4.2e-04 M


In [21]:
# --- Acids to compare ---
acids = ["HC2F3O2", "HC2FH2O2", "HC2F2HO2"]

# --- Get the ranking ---
ranking = ab.rank_acids_by_inductive_effect(acids)

print("The order of increasing acid strength is:")
print(ranking)

The order of increasing acid strength is:
HC2FH2O2 < HC2F2HO2 < HC2F3O2


# Moles based on volume and moles

In [22]:
# --- Given values ---
molarity_hcl = 0.200  # M
volume_hcl_mL = 15.0  # mL

# The mole ratio from the balanced equation (1 mol LiOH / 1 mol HCl) is 1
mole_ratio = 1

# --- Calculation ---
# 1. Calculate moles of HCl used
moles_hcl = molarity_hcl * (volume_hcl_mL / 1000)

# 2. Use the mole ratio to find moles of LiOH
moles_lih = moles_hcl * mole_ratio

print(f"The moles of LiOH in the sample are {moles_lih:.5f} mol")

The moles of LiOH in the sample are 0.00300 mol


# Calculate pH of a buffer

In [23]:
# --- Given values ---
pKa_carbonic_acid = 6.37
conc_acid = 0.20   # M of H₂CO₃
conc_base = 0.10   # M of NaHCO₃

# --- Calculation ---
pH = ab.calculate_buffer_ph_from_pka(pKa_carbonic_acid, conc_acid, conc_base)

print(f"The pH of the buffer is {pH:.2f}")

The pH of the buffer is 6.07


In [24]:

# --- Given values ---
# Define compounds to get their molar mass
acetic_acid = Compound("HC2H3O2") # Formula for acetic acid
sodium_acetate = Compound("NaC2H3O2") # Formula for sodium acetate

# Set the mass for each to calculate moles
acetic_acid.set_mass(15.01)
sodium_acetate.set_mass(20.51)

# Set the volume to calculate concentration
solution_volume = 1.00 # L
acetic_acid.set_volume(solution_volume)
sodium_acetate.set_volume(solution_volume)

# Ka value
Ka_acetic_acid = 1.80e-5

# --- Calculation ---
pH = ab.calculate_buffer_pH(
    Ka=Ka_acetic_acid,
    conc_acid=acetic_acid.molarity_M,
    conc_base=sodium_acetate.molarity_M
)

print(f"The pH of the buffer solution is {pH:.2f}")

The pH of the buffer solution is 4.74


In [25]:
# --- 1. Define the components and their properties ---
# Acetic Acid (the weak acid)
acetic_acid = Compound("CH3COOH")
acetic_acid.set_mass(37.95)

# Sodium Acetate (the conjugate base)
sodium_acetate = Compound("NaC2H3O2")
sodium_acetate.set_mass(61.11)

# Set the solution volume for both to calculate molarity
solution_volume_L = 0.500  # 500.0 mL
acetic_acid.set_volume(solution_volume_L)
sodium_acetate.set_volume(solution_volume_L)

# Ka value from the problem
Ka_acetic_acid = 1.8e-5

# --- 2. Use your function to calculate the buffer pH ---
buffer_ph = ab.calculate_buffer_pH(
    Ka=Ka_acetic_acid,
    conc_acid=acetic_acid.molarity_M,
    conc_base=sodium_acetate.molarity_M
)

print(f"The calculated pH of the buffer solution is {buffer_ph:.2f}")

The calculated pH of the buffer solution is 4.82


In [26]:
# --- 1. Define the components and their properties ---
# Acetic Acid (the weak acid)
acetic_acid = Compound("CH3COOH")
acetic_acid.set_mass(37.95)

# Sodium Acetate (the conjugate base)
sodium_acetate = Compound("NaC2H3O2")
sodium_acetate.set_mass(61.11)

# Set the solution volume for both to calculate molarity
solution_volume_L = 0.500  # 500.0 mL
acetic_acid.set_volume(solution_volume_L)
sodium_acetate.set_volume(solution_volume_L)

# Ka value from the problem
Ka_acetic_acid = 1.8e-5

# --- 2. Use your function to calculate the buffer pH ---
buffer_ph = ab.calculate_buffer_pH(
    Ka=Ka_acetic_acid,
    conc_acid=acetic_acid.molarity_M,
    conc_base=sodium_acetate.molarity_M
)

print(f"The calculated pH of the buffer solution is {buffer_ph:.2f}")

The calculated pH of the buffer solution is 4.82


In [27]:
# --- Given values from the problem ---
ka_lactic_acid = 1.4e-4
mass_acid = 41.80
fm_acid = 90.08
mass_base = 75.30
fm_base = 112.06

# --- Use the new function to calculate the pH ---
buffer_ph = ab.calculate_buffer_ph_from_mass(
    Ka=ka_lactic_acid,
    mass_acid_g=mass_acid,
    fm_acid=fm_acid,
    mass_base_g=mass_base,
    fm_base=fm_base
)

print(f"The calculated pH of the buffer solution is {buffer_ph:.2f}")

The calculated pH of the buffer solution is 4.01


# Calculate pka from buffer, ph and moles

In [28]:
# --- Given values ---
pH_buffer = 4.41
conc_acid = 0.32  # M
conc_base = 0.14  # M

# --- Calculation ---
pKa_acid = ab.calculate_pka_from_buffer_ph(pH_buffer, conc_acid, conc_base)

print(f"The pKa of the acid is {pKa_acid:.2f}")

The pKa of the acid is 4.77


In [29]:
# --- Given values ---
pH_buffer = 3.95
conc_acid = 0.50  # M
conc_base = 0.25  # M

# --- Calculation ---
pKa_acid = ab.calculate_pka_from_buffer_ph(pH_buffer, conc_acid, conc_base)

print(f"The pKa of the acid is {pKa_acid:.2f}")

The pKa of the acid is 4.25


# FiRE table and concentration of weak acid and conjugate base after acid addition

In [30]:
import acids_and_bases as ab

# --- Part 1: Directly Answer the Question (Calculate Initial Moles) ---
# This is what you need to solve the problem in the image.

# Given values from the image
vol_buffer_L = 0.850  # 850. mL
conc_acid_M = 1.50   # M of HClO
conc_base_M = 1.65   # M of ClO-

vol_added_L = 0.050  # 50.0 mL
conc_added_M = 2.50   # M of KOH

# Calculate the initial moles of each component
initial_moles_acid = vol_buffer_L * conc_acid_M
initial_moles_base = vol_buffer_L * conc_base_M
initial_moles_added = vol_added_L * conc_added_M

print("--- Initial Moles for FIRE Table ---")
print(f"Initial Moles of HClO (Acid): {initial_moles_acid:.3f} mol")
print(f"Initial Moles of ClO⁻ (Base): {initial_moles_base:.3f} mol")
print(f"Initial Moles of OH⁻ (Added): {initial_moles_added:.3f} mol")
print("-" * 35)

--- Initial Moles for FIRE Table ---
Initial Moles of HClO (Acid): 1.275 mol
Initial Moles of ClO⁻ (Base): 1.402 mol
Initial Moles of OH⁻ (Added): 0.125 mol
-----------------------------------


# Compare buffer capacity

In [ ]:
# Define the buffers from the question
buffer_options = [
    {'conc_acid': 0.067, 'conc_base': 0.071},
    {'conc_acid': 0.867, 'conc_base': 0.158},
    {'conc_acid': 0.324, 'conc_base': 0.173},
    {'conc_acid': 0.625, 'conc_base': 0.635}
]

# Get the analysis
analysis_report = ab.compare_buffer_capacities(buffer_options)
print(analysis_report)

--- Buffer Capacity Analysis ---

| Buffer | Total Conc. (M) | Ratio [Base]/[Acid] |
|:-------|:---------------:|:--------------------:|
|   #1   |      0.138      |        1.06         |
|   #2   |      1.025      |        0.18         |
|   #3   |      0.497      |        0.53         |
|   #4   |      1.260      |        1.02         |

--- Conclusion ---
Buffer #4 (0.625 M Acid / 0.635 M Base) has the greatest buffering capacity.
This is because it has the highest concentration of buffer components while maintaining a ratio close to 1.


In [37]:
# Define the buffers from the question
buffer_options = [
    {'conc_acid': 0.275, 'conc_base': 0.752},
    {'conc_acid': 0.855, 'conc_base': 0.322},
    {'conc_acid': 0.760, 'conc_base': 0.722},
    {'conc_acid': 0.245, 'conc_base': 0.322}
]

# Get the analysis
analysis_report = ab.compare_buffer_capacities(buffer_options)
print(analysis_report)

--- Buffer Capacity Analysis ---

| Buffer | Total Conc. (M) | Ratio [Base]/[Acid] |
|:-------|:---------------:|:--------------------:|
|   #1   |      1.027      |        2.73         |
|   #2   |      1.177      |        0.38         |
|   #3   |      1.482      |        0.95         |
|   #4   |      0.567      |        1.31         |

--- Conclusion ---
Buffer #3 (0.76 M Acid / 0.722 M Base) has the greatest buffering capacity.
This is because it has the highest concentration of buffer components while maintaining a ratio close to 1.


# pH at halfway to the first equivalence point

In [32]:
# Ka values for carbonic acid from the problem
ka_values_h2co3 = [4.30e-7, 5.60e-11]

# Calculate the pH at the first half-equivalence point
ph_at_halfway = ab.titration_ph_at_half_equivalence(
    Ka_values=ka_values_h2co3,
    point_number=1
)

print(f"The pH at the halfway point to the first equivalence point is: {ph_at_halfway:.2f}")

The pH at the halfway point to the first equivalence point is: 6.37


# Conjugate base is most suitable to prepare a buffer 

In [33]:
# --- Given values from the question ---
target_ph = 4.50
acids = {
    "acetic acid": 1.8e-5,
    "tris": 8.5e-9,
    "ammonium chloride": 5.6e-10,
    "boric acid": 5.8e-10
}

best_acid = None
smallest_difference = float('inf')

print(f"Finding the best acid for a buffer with pH = {target_ph}\n")

for name, ka in acids.items():
    pka = -math.log10(ka)
    difference = abs(target_ph - pka)
    
    print(f"- {name:<20} | pKa = {pka:.2f} | Difference = {difference:.2f}")

    if difference < smallest_difference:
        smallest_difference = difference
        best_acid = name

print(f"\nConclusion: {best_acid.capitalize()} is the most suitable choice.")

Finding the best acid for a buffer with pH = 4.5

- acetic acid          | pKa = 4.74 | Difference = 0.24
- tris                 | pKa = 8.07 | Difference = 3.57
- ammonium chloride    | pKa = 9.25 | Difference = 4.75
- boric acid           | pKa = 9.24 | Difference = 4.74

Conclusion: Acetic acid is the most suitable choice.


# Lewis Acid/Base Classification

In [34]:
# List of formulas
species_list = ["Au", "CO2", "BF3", "H2O", "Au"]

# Create Compound objects from the formulas
compound_list = [Compound(s) for s in species_list]
# Manually set the charge on the first Au to make it Au+
compound_list[0].charge = 1

print("--- Lewis Acid/Base Classification ---")
for compound_obj in compound_list:
    # Get the formula string for printing
    name_to_print = compound_obj.input_formula
    if compound_obj.charge is not None:
        sign = "+" if compound_obj.charge > 0 else "-"
        name_to_print += str(abs(compound_obj.charge)) + sign


    classification = ab.classify_lewis_acid_base(compound_obj)
    print(f"{name_to_print:<5} -> {classification}")

--- Lewis Acid/Base Classification ---
Au1+  -> Lewis Acid (cation)
CO2   -> Lewis Acid (central atom can accept e- pair)
BF3   -> Lewis Acid (incomplete octet on B)
H2O   -> Lewis Base (has lone pair(s) to donate)
Au    -> Neither (neutral atom)


In [35]:
compound = Compound("BF3")
classification = ab.classify_lewis_acid_base(compound)
print(f"{compound.input_formula} -> {classification}")

BF3 -> Lewis Acid (incomplete octet on B)


# Calculate buffer ph after addition from moles

In [36]:
# --- Given values from the problem ---
moles_HIO = 0.680
moles_KIO = 0.500
moles_KOH = 0.200
Ka_HIO = 3e-11

# --- Use the new function to calculate the final pH ---
final_ph = ab.calculate_buffer_ph_after_addition_from_moles(
    initial_moles_acid=moles_HIO,
    initial_moles_base=moles_KIO,
    added_moles=moles_KOH,
    added_is_acid=False,  # We added a base (KOH)
    Ka=Ka_HIO
)

print(f"The final pH of the buffer is: {final_ph:.2f}")

The final pH of the buffer is: 10.69


In [38]:
# --- Given values from the problem ---
moles_HCN = 0.375
moles_NaCN = 0.375
moles_HNO3 = 0.100
Ka_HCN = 6.2e-10

# --- Use your function to calculate the final pH ---
final_ph = ab.calculate_buffer_ph_after_addition_from_moles(
    initial_moles_acid=moles_HCN,
    initial_moles_base=moles_NaCN,
    added_moles=moles_HNO3,
    added_is_acid=True,  # We added an acid (HNO3)
    Ka=Ka_HCN
)

print(f"The final pH of the buffer is: {final_ph:.2f}")

The final pH of the buffer is: 8.97


## The Principle of Buffer Selection
![image.png](images/buffer-1.png)

A buffer solution is most effective at resisting pH changes when the concentrations of the weak acid and its conjugate base are equal. According to the Henderson-Hasselbalch equation, when these concentrations are equal, the pH of the buffer is equal to the pKₐ of the weak acid.

pH=pKa​ (when [Acid] = [Base])

Therefore, to create an effective buffer at a specific pH, you should choose a conjugate acid-base pair where the acid's pKₐ is as close as possible to the target pH.

### Analysis of the Options

We need to find the buffer system whose pKₐ is closest to the target pH of 6.8.

    Sodium hydrogen carbonate (pKₐ = 10.3): Too far from 6.8.

    Acetic acid (pKₐ = 4.8): Too far from 6.8.

    Hypochlorous acid (pKₐ = 7.5): This is the closest pKₐ to the target pH of 6.8.

    Hydrocyanic acid (pKₐ = 9.3): Too far from 6.8.

By choosing the hypochlorous acid system, a chemist can create an effective buffer at pH 6.8 by adjusting the ratio of the conjugate base to the acid.